homeassistant/components/godice/sensor.py

```diff
@@ -0,0 +1,136 @@
+"""Sensors for GoDice device."""
+
+import logging
+
+import godice
+
+from homeassistant import const
+from homeassistant.components.sensor import (
+    RestoreSensor,
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DATA_DEVICE, DATA_DEVICE_INFO, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+ROLLED_NUMBER_SENSOR_DESCR = SensorEntityDescription(
+    key="rolled_number", state_class=SensorStateClass.MEASUREMENT
+)
+COLOR_SENSOR_DESCR = SensorEntityDescription(
+    key="color",
+    device_class=SensorDeviceClass.ENUM,
+    options=[color.name for color in godice.Color],
+)
+BATTERY_SENSOR_DESCR = SensorEntityDescription(
+    key="battery_level",
+    device_class=SensorDeviceClass.BATTERY,
+    native_unit_of_measurement=const.PERCENTAGE,
+    state_class=SensorStateClass.MEASUREMENT,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Setups Dice sensors."""
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    device = data[DATA_DEVICE]
+    device_info = data[DATA_DEVICE_INFO]
+
+    entry_ctors = [
+        DiceColorSensor,
+        BatteryLevelSensor,
+        DiceNumberSensor,
+    ]
+
+    entries = [ctor(device_info, device) for ctor in entry_ctors]
+    async_add_entities(entries)
+
+
+class BaseSensor(SensorEntity):
+    """Base for concrete sensors."""
+
+    _attr_should_poll = False
+    _attr_has_entity_name = True
+
+    def __init__(
+        self, entity_description: SensorEntityDescription, device_info: DeviceInfo
+    ) -> None:
+        """Set default values."""
+        self.entity_description = entity_description
+        self._attr_device_info = device_info
+        self._attr_unique_id = (
+            f"{device_info[const.CONF_NAME]}_{self.entity_description.key}"
+        )
+        self._attr_translation_key = entity_description.key
```

Comment:
Related to the other suggestion about DeviceInfo, I would include it here in the BaseSensor and then pass the ConfigEntry to the constructor

```suggestion
class BaseSensor(SensorEntity):
    """Base for concrete sensors."""

    _attr_should_poll = False
    _attr_has_entity_name = True
    
    dice: godice.Dice

    def __init__(
        self, entity_description: SensorEntityDescription, entry: ConfigEntry, dice: godice.Dice
    ) -> None:
        """Set default values."""
        self.dice = dice
        self.entity_description = entity_description
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, device_name)},
            name=entry.data[CONF_NAME],
            manufacturer="Particula",
            model="GoDice",
        ),
        self._attr_unique_id = (
            f"{device_info[const.CONF_NAME]}_{self.entity_description.key}"
        )
        self._attr_translation_key = entity_description.key
```

Reply:
moved DeviceInfo var declaration to sensors. Passing DeviceInfo to sensor classes instead of ConfigEntry since only DeviceInfo is needed there

Reply:
I would personally still create the DeviceInfo in the BaseSensor class. In that case everything related to the BaseSensor is defined in that class, just like how the _attr_unique_id is defined in the constructor which makes it easier to navigate. Also, it reduces the size of your async_setup_entry:
```python
async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Setups Dice sensors."""
    device = hass.data[DOMAIN][config_entry.entry_id][DATA_DEVICE]

    sensor_create_data = [
        (DiceColorSensor, COLOR_SENSOR_DESCR),
        (BatteryLevelSensor, BATTERY_SENSOR_DESCR),
        (DiceNumberSensor, ROLLED_NUMBER_SENSOR_DESCR),
    ]

    async_add_entities(
        [
            entity_class(entry, entity_description, device)
            for entity_class, entity_description in sensor_create_data
        ]
    )
```
```python
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, device_name)},
            name=entry.data[CONF_NAME],
            manufacturer="Particula",
            model="GoDice",
        ),
```

Reply:
creating `DeviceInfo` once and passing the same object for all sensors better demonstrates the fact that all sensors belong to the same device, in contrast of creating `DeviceInfo` per each Sensor

Reply:
That's also true, advantages to both options. I personally prefer to see DeviceInfo in the Entity class but you can also leave it like this if you prefer this